### PR TITLE
feat(purrr): implement remaining PR events

### PR DESCRIPTION
**Pull request by daabr**
Un/assign reviewer, PR edited (title, description, base branch), PR synchronized (head branch).

Not tested manually yet (only unit testing) - this is still blocked by ENG-1913.

Refs: INT-152

### CODE REVIEW

File: purrr/github_pr.py [hunk 1]
```diff
@@ -335,36 +348,79 @@ def _on_pr_unassigned(data) -> None:
...
+    if not channel:
         return
 
-    pass  # TODO: Implement this function.
+    assignee, sender = data.assignee, data.sender
+    _on_pr_review_request_removed_person(assignee, sender, channel, "assignee")
 
 
-def _on_pr_edited(data) -> None:
+def _on_pr_edited(action: str, pr, changes, sender) -> None:
```
Comment (mario99logic) [inline purrr/github_pr.py:359]: Why not add type hints to the other parameters

File: purrr/github_pr.py [hunk 2]
```diff
@@ -335,36 +348,79 @@ def _on_pr_unassigned(data) -> None:
...
+    channel = _lookup_channel(pr, action)
+    if not channel:
         return
 
-    pass  # TODO: Implement this function.
+    # PR base branch was changed.
+    if "base" in changes:
+        msg = "{} changed the base branch from "
+        msg += "`{changes.base.ref}` to `{pr.base.ref}`"
+        slack_helper.mention_in_message(channel, sender, msg)
```
Comment (mario99logic) [inline purrr/github_pr.py:377]: Does base branch here mean the main branch and if so why are we checking if the main branch changed on a pr edited

**Comment by mario99logic** [comment]
The tests effectively cover different scenarios for reviewers and assignees. Is there edge cases such a invalid user IDs?

**Review by mario99logic — APPROVED** [review]

File: purrr/github_pr_test.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+def mock_slack_user_id(mocker):
+    import users
+
+    return mocker.patch.object(users, "github_username_to_slack_user_id", autospec=True)
+
+
+FakeGithubUser = collections.namedtuple("FakeGithubUser", ["login"])
+
+
+class TestOnPRReviewRequestedPerson:
```
Comment (mario99logic) [inline purrr/github_pr_test.py:25]: The tests effectively cover different scenarios for reviewers and assignees. Is there edge cases such a invalid user IDs?

File: purrr/github_pr.py (hunk 1 above)
Comment (daabr) [reply purrr/github_pr.py:359]: There's no good definition of the structure of these objects, so no type hint is the most succinct way to show that. Another alternative is to specify an `Any` type, but that's redundant.

File: purrr/github_pr.py (hunk 2 above)
Comment (daabr) [reply purrr/github_pr.py:377]: `main` is a common base branch, but the PR author can use any branch they want, e.g. the head branch of another PR. A PR with a non-`main` base branch is called a "stacked PR" (i.e. it's "stacked" on top of another PR). You can read more about this here for example: https://axolo.co/blog/p/managing-stacked-pr

File: purrr/github_pr_test.py (hunk 3 above)
Comment (daabr) [reply purrr/github_pr_test.py:25]: Good question.
- The function under test (`_on_pr_review_requested_person`) receives its input from a GitHub event (that triggered the workflow)
- AK guarantees that this event is really from GitHub and not fake (we check signatures and discard fake payloads)
- So the `reviewer` and `sender` inputs are guaranteed to be valid GitHub user objects, specifically they contain a real `login` string field
- It **is** possible that a real GitHub user doesn't have a Slack account (e.g. external contributor, not an employee of the org)
- The function under test **does** handle this possibility, but it simply means returning without doing some things in Slack, so it's not an interesting test IMO

**Review by pashafateev — APPROVED** [review]
difficult to put this PR in perspective of the larger Purrr project, but nothing jumped out at me.
